### PR TITLE
Integrate PowerShell cloud shell with the PowerShell kernel

### PR DESCRIPTION
**Pull request by daxian-dbw**
Add `Enter-AzShell` cmdlet to `Microsoft.Dotnet.Interactive.PowerShell` module to integrate Azure cloud shell with the PowerShell kernel.

Running `Enter-AzShell` will authenticate you to Azure and establish a PowerShell cloud shell session for you. After that, all subsequent code submitted from the cells are executed in that cloud shell session.
When you are done with the cloud shell, just type and run `exit` from a cell to end the session, and after that, all subsequent code submitted from the cells will again be executed using the local PowerShell kernel.

/cc @SteveL-MSFT 

Mar-18-2020 00-39-31

### CODE REVIEW

**Comment by daxian-dbw** [comment]
@SteveL-MSFT Can you please review this PR when you get a chance? You are most familiar with the involved changes.

**Comment by daxian-dbw** [comment]
We could do something similar to light up `Enter-PSSession` in PowerShell kernel -- run submitted code in a remote PSSession.
And we could abstract out the remote code run interface, to cover both the `Enter-PSSession` and `Enter-AzShell` scenarios.
See  https://github.com/dotnet/interactive/issues/290 for details.

File: Microsoft.DotNet.Interactive.PowerShell/Commands/EnterAzShellCommand.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        [Parameter(ParameterSetName = DefaultSetName)]
+        public SwitchParameter Reset;
+
+        /// <summary>
+        /// ProcessRecord override.
+        /// </summary>
+        protected override void ProcessRecord()
+        {
+            var context = KernelInvocationContext.Current;
+            if (context?.CurrentKernel is PowerShellKernel psKernel)
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.PowerShell/Commands/EnterAzShellCommand.cs]: This API is confusing, I know, but this should probably be `HandlingKernel`, not `CurrentKernel`.

File: Microsoft.DotNet.Interactive.PowerShell/Host/VTColorUtils.cs [hunk 2]
```diff
@@ -9,6 +9,7 @@ internal static class VTColorUtils
     {
         internal const string EscapeCharacters = "\x1b[";
         internal const string ResetColor = "\x1b[0m";
+        internal const string BoldColor = "\x1b[1m";
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.PowerShell/Host/VTColorUtils.cs]: We've added a reference to `System.CommandLine.Rendering` so we might want to refactor to use that for VT. I'm happy to help with the the refactor, as I'd love feedback on the APIs.

File: Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs [hunk 3]
```diff
@@ -150,72 +156,120 @@ public override bool TryGetVariable(string name, out object value)
...
+            }
 
-            var completionList = completion.CompletionMatches.Select(c => new CompletionItem(
-                displayText: c.CompletionText,
-                kind: c.ResultType.ToString(),
-                documentation: c.ToolTip
-            ));
+            if (shouldDispose)
+            {
+                AzShell.Dispose();
```
Comment (jonsequitur) [inline Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs]: You might also want to do `base.RegisterForDisposal` for `AzShell` when it's initialized. For example, do we care if the `PowerShellKernel` gets disposed without someone having done an explicit `exit`?

File: Microsoft.DotNet.Interactive.PowerShell/Commands/EnterAzShellCommand.cs (hunk 1 above)
Comment (daxian-dbw) [reply Microsoft.DotNet.Interactive.PowerShell/Commands/EnterAzShellCommand.cs]: Adding comment to `HandlingKernel` and `CurrentKernel` properties should be sufficient.
`CurrentKernel` is the one the notebook running against initially, `HandlingKernel` could be changed when using `#!pwsh` for example, is this understanding correct?

File: Microsoft.DotNet.Interactive.PowerShell/Host/VTColorUtils.cs (hunk 2 above)
Comment (daxian-dbw) [reply Microsoft.DotNet.Interactive.PowerShell/Host/VTColorUtils.cs]: Would love to refactor and use that library! Opened #299 to track the work.

File: Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs (hunk 3 above)
Comment (daxian-dbw) [reply Microsoft.DotNet.Interactive.PowerShell/PowerShellKernel.cs]: Good point. Will address this.

**Pull request merged**